File: 1-1000/0945-minimum-increment-to-make-array-unique/0945-minimum-increment-to-make-array-unique.cpp
class Solution {
public:
    int minIncrementForUnique(vector<int>& nums) {
        int max_val = 0;
        int size = nums.size();
        for(int num:nums){
            max_val = max(max_val,num); 
        }
        vector<int>count(size+max_val,0);
        long long int answer =0;
        //int max_min_num = 100005; //use after all min_heap elements are used;
        //priority_queue<int,vector<int>,greater<int> > min_heap;
        //this min_heap in not required for more optimizations;

        for(int i=0;i<nums.size();i++){
           count[nums[i]]++;
        }
        for(int i=0;i<count.size();i++){
            if(count[i]<=1){
                continue;
            }
            int dup = count[i]-1;
            count[i+1]+=dup;
            count[i]=1;
            answer+=dup;
        }
        return answer;
    }
};

File: 1-1000/0945-minimum-increment-to-make-array-unique/0945-minimum-increment-to-make-array-unique.java
class Solution {
    public int minIncrementForUnique(int[] nums) {
        int max = nums[0];
        for(int i=0;i<nums.length;i++){
            if(nums[i] > max)
            max = nums[i];
        }

        int arr[] = new int[nums.length + max + 1];

        for(int i=0;i<nums.length;i++){
            arr[nums[i]]+=1;
        }

        int count = 0;
        for(int i=0;i<arr.length - 1;i++){
            if(arr[i] == 0)
            continue;

            int d = arr[i] - 1;
            arr[i + 1] += d;
            count+=d;
        }
        return count;
    }

}